### PR TITLE
fix: Fixes x509 cert issue by passing trusted certs via CM for BitBucketServer and GitLab

### LINKED ISSUE #9607: ApplicationSet controller receives x509: certificate signed by unknown authority, when using SCM, PR or Git generator with private bitbucket server
Slack conversation thread : https://cloud-native.slack.com/archives/C01U45M2SVB/p1654569552211179
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**
Receive `x509: certificate signed by unknown authority` when application set controller with an SCM generator attempts to connect to a private bitbucket server.

**To Reproduce**

1) Follow installation for release v2.4.0-rc4 under https://github.com/argoproj/argo-cd/releases
2) Follow steps to setup TLS certificates as documented https://argo-cd.readthedocs.io/en/release-1.8/user-guide/private-repositories/#self-signed-untrusted-tls-certificates
3) Create an ApplicationSet CRD with an SCM generator for the Bitbucket Server SCM. The bitbucket server needs to be a private repository with TLS connectivity via self signed certificates which should have been registered in the above step

**Expected behavior**

Expected the application set controller to be able to communicate with bitbucket server API's to list repositories and automate ArgoCD application generation. 

**Screenshots**

**Version**
```
{
    "Version": "v2.4.0-rc2+cd5b2af",
    "BuildDate": "2022-05-18T12:42:45Z",
    "GitCommit": "cd5b2af358f3574ea151fr8c24b1d52dafe59c85",
    "GitTreeState": "clean",
    "GoVersion": "go1.18.2",
    "Compiler": "gc",
    "Platform": "linux/amd64",
    "KustomizeVersion": "v4.4.1 2021-11-11T23:36:27Z",
    "HelmVersion": "v3.8.1+g5cb9af4",
    "KubectlVersion": "v0.23.1",
    "JsonnetVersion": "v0.18.0"
}
```
**Logs**

- applicationset-controller logs

```
time="2022-06-08T00:45:28Z" level=info msg="Starting configmap/secret informers"
time="2022-06-08T00:45:28Z" level=info msg="Configmap/secret informer synced"
time="2022-06-08T00:45:28Z" level=info msg="Starting webhook server"
time="2022-06-08T00:45:28Z" level=info msg="Starting manager"
time="2022-06-08T00:45:29Z" level=error msg="error generating params" error="error listing repos: error listing repositories for: Get \"https://<myBitBucketServerHost>/rest/api/1.0/projects/<myProject>/repos?limit=100\": x509: certificate signed by unknown authority" generator="&{0xc000c6c1e0 <nil>}"
```

- Volume and Volume Mounts of the applicationset-controller
```
 Mounts:
      /app/config/ssh from ssh-known-hosts (rw)
      /app/config/tls from tls-certs (rw)
      /tmp from tmp (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from argocd-applicationset-controller-token-59cjk (ro)


Volumes:
  tls-certs:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      argocd-tls-certs-cm
    Optional:  false
  ssh-known-hosts:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      argocd-ssh-known-hosts-cm
    Optional:  false
```

### PR DESCRIPTION
**Pull request by appiepollo14**
Fixes https://github.com/argoproj/argo-cd/issues/9607 for BitBucketServer by allowing to pass a CM holding trusted certificates. Refactors implementation for GitLab. As written in https://github.com/argoproj/argo-cd/issues/9607#issuecomment-2212500925 the implementation is done by adding the option to pass trusted certs via a CM. In this way, we can reference the CM holding ArgoCD's trusted certs. This solves the issue that GitOps Operator users, can't mount volumes to the appSet controller which was a workaround for passing trusted CA's for them.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Comment by appiepollo14** [comment]
@agaudreault would you take a look?

**Comment by appiepollo14** [comment]
How to get some movement on this PR? @ishitasequeira @agaudreault ?

**Comment by ishitasequeira** [comment]
> How to get some movement on this PR? @ishitasequeira @agaudreault ?

@appiepollo14 I'll add this to my current sprint

File: applicationset/generators/scm_provider.go [hunk 1]
```diff
@@ -276,6 +292,25 @@ func (g *SCMProviderGenerator) getSecretRef(ctx context.Context, ref *argoprojio
 	return string(tokenBytes), nil
 }
 
+func (g *SCMProviderGenerator) getConfigMapData(ctx context.Context, ref *argoprojiov1alpha1.ConfigMapKeyRef, namespace string) ([]byte, error) {
```
Comment (agaudreault) [inline applicationset/generators/scm_provider.go:295]: Duplicated code in the 2 providers. Can you move to the utils package? Maybe a `kubernetes.go` or `client.go`. `getSecretRef` should probably be moved as well.

File: pkg/apis/application/v1alpha1/applicationset_types.go [hunk 2]
```diff
@@ -528,6 +536,10 @@ type SCMProviderGeneratorBitbucketServer struct {
 	AllBranches bool `json:"allBranches,omitempty" protobuf:"varint,4,opt,name=allBranches"`
 	// Credentials for AccessToken (Bearer auth)
 	BearerToken *BearerTokenBitbucket `json:"bearerToken,omitempty" protobuf:"bytes,5,opt,name=bearerToken"`
+	// Allow self-signed TLS / Certificates; default: false
+	Insecure bool `json:"insecure,omitempty" protobuf:"varint,6,opt,name=insecure"`
+	// ConfigMap key holding the trusted certificates
+	CAConfigMapKeyRef *ConfigMapKeyRef `json:"caConfigMapKeyRef,omitempty" protobuf:"bytes,7,opt,name=caConfigMapKeyRef"`
```
Comment (agaudreault) [inline pkg/apis/application/v1alpha1/applicationset_types.go:542]: Seems repetitive to have `CAConfigMapKeyRef.ConfigMapName`. I don't think it is necessary to reflect they type of the object in the property name. `CARef.ConfigMapName` aligns with existing terminology used with `SecretRef`.

**Review by agaudreault — CHANGES_REQUESTED** [review]
Few nitpicks. Nothing major 👍

File: pkg/apis/application/v1alpha1/applicationset_types.go (hunk 2 above)
Comment (appiepollo14) [reply pkg/apis/application/v1alpha1/applicationset_types.go:542]: Agreed!

File: applicationset/generators/scm_provider.go (hunk 1 above)
Comment (appiepollo14) [reply applicationset/generators/scm_provider.go]: Done

File: applicationset/generators/pull_request.go [hunk 3]
```diff
@@ -202,33 +218,52 @@ func (g *PullRequestGenerator) github(ctx context.Context, cfg *argoprojiov1alph
...
 	return pullrequest.NewGithubService(ctx, token, cfg.API, cfg.Owner, cfg.Repo, cfg.Labels)
 }
 
-// getSecretRef gets the value of the key for the specified Secret resource.
-func (g *PullRequestGenerator) getSecretRef(ctx context.Context, ref *argoprojiov1alpha1.SecretRef, namespace string) (string, error) {
-	if ref == nil {
-		return "", nil
-	}
+// // getSecretRef gets the value of the key for the specified Secret resource.
+// func (g *PullRequestGenerator) getSecretRef(ctx context.Context, ref *argoprojiov1alpha1.SecretRef, namespace string) (string, error) {
```
Comment (ishitasequeira) [inline applicationset/generators/pull_request.go:229]: As this is moved to the utils file, can we remove the commented code?

File: applicationset/generators/pull_request_test.go [hunk 4]
```diff
@@ -224,70 +222,137 @@ func TestPullRequestGithubGenerateParams(t *testing.T) {
...
-			"my-token": []byte("secret"),
-		},
-	}
-	gen := &PullRequestGenerator{client: fake.NewClientBuilder().WithObjects(secret).Build()}
-	ctx := context.Background()
+// func TestPullRequestGetSecretRef(t *testing.T) {
+// 	secret := &corev1.Secret{
+// 		ObjectMeta: metav1.ObjectMeta{Name: "test-secret", Namespace: "test"},
+// 		Data: map[string][]byte{
+// 			"my-token": []byte("secret"),
```
Comment (ishitasequeira) [inline applicationset/generators/pull_request_test.go:229]: Same here

File: applicationset/generators/scm_provider.go [hunk 5]
```diff
@@ -253,28 +268,47 @@ func (g *SCMProviderGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha
...
-	if !ok {
-		return "", fmt.Errorf("key %q in secret %s/%s not found", ref.Key, namespace, ref.SecretName)
-	}
-	return string(tokenBytes), nil
-}
+// func (g *SCMProviderGenerator) getSecretRef(ctx context.Context, ref *argoprojiov1alpha1.SecretRef, namespace string) (string, error) {
+// 	if ref == nil {
+// 		return "", nil
+// 	}
+
```
Comment (ishitasequeira) [inline applicationset/generators/scm_provider.go:275]: Same here

**Review by ishitasequeira — CHANGES_REQUESTED** [review]
Thanks @appiepollo14 !! Overall code looks good!! 

I think just removing of commented code is pending. Will wait for that to merge the PR.

**Comment by appiepollo14** [comment]
@ishitasequeira @agaudreault thanks for the reviews, have updatet the code based on your comments.

**Review by ishitasequeira — APPROVED** [review]
LGTM!! Thanks @appiepollo14!

**Comment by appiepollo14** [comment]
@ishitasequeira can this be backported to 2.11..?

**Comment by agaudreault** [comment]
@appiepollo14 It would also need to be backported to 2.12 as release is almost out. This PR contain a change in the CRDs so I dont think it is something that should be backported.

**Comment by appiepollo14** [comment]
@agaudreault true. It's up to you of course. My worries are in the fact that it then would time a rather long time before it reaches a gitops operator release. Hence why I ask to backport.

**Pull request merged**